### ISSUE TITLE
Unable to Modify ulimit Within /etc/docker/daemon.json

### ISSUE THREAD

**Issue by AdamBouhmad** [docker/docker#29385]
**Description**

Hey all! I Recently found an odd bug with the Docker Daemon through my adventures of hardening a container. I want to modify the values for ulimit via the daemon.json file in accordance with the [CIS_Docker 1.12 benchmark](https://benchmarks.cisecurity.org/tools2/docker/CIS_Docker_1.12.0_Benchmark_v1.0.0.pdf), and I quickly referred to the [docker docs](https://docs.docker.com/engine/reference/commandline/dockerd/#/linux-configuration-file), however it seems through my search, I came up empty on a documented way to do this. I attempted to follow [#25566](https://github.com/docker/docker/pull/25566) as a guide, however values that I input to the "default-ulimits" field return errors, and thus the docker daemon is unable to be started. 

**Steps to reproduce the issue:**
1. If it's not already available, create the following file:
```
/etc/docker/daemon.json
```
2. Add in the following line: 
```
{"default-ulimits": {"nofile":"100:200"}}
```
3. Restart the Docker daemon.

**Describe the results you received:**
level=fatal msg="unable to configure the Docker daemon with file /etc/docker/daemon.json: the following directives don't match any configuration option: nofile\n"

**Describe the results you expected:**
Docker daemon should be able to start, and pass the --default-ulimit as an argument with the respective ulimits. 

**Additional information you deem important (e.g. issue happens only occasionally):**
This runs cleanly through the command line for just nproc, and same for nofile: 
```
dockerd --default-ulimit nproc=1024:2408 --default-ulimit nofile=100:200
```
**Output of `docker version`:**

```
Docker version 1.12.3, build 6b644ec
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
AWS

**Extra:** In the case that it's not a bug and simply lack of documentation, I would be more than willing to help document the use cases for ulimit, as currently there are no examples of modifying ulimit within daemon.json. 

Thanks in advance!

**Comment by thaJeztah** [comment]
Thanks for reporting; this issue was already reported in https://github.com/docker/docker/issues/22309, and will be fixed in docker 1.13 through https://github.com/docker/docker/pull/25566. Given that it's a [simple fix](https://github.com/docker/docker/pull/25566/files#diff-a348195a1b645d5b477946b1efae728bR48), perhaps there's still a chance to get this into a patch release, but no promises :smile:

For now; if your system is using systemd as an init system, use a drop-in file to set the `--default-ulimit` flag; https://docs.docker.com/engine/admin/systemd/#/custom-docker-daemon-options

If your system uses upstart or sysvinit, add the flag to the `DOCKER_OPTS` variable, following the instructions here; https://docs.docker.com/engine/admin/#/configuring-docker

/cc @vieux

**Comment by thaJeztah** [comment]
Looking at the status for the next patch release, I think we won't be able to get this into that release; given that there are other ways to configure this (using the options mentioned above), I'm going to go ahead and close this issue, as it will be fixed for 1.13.
